lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java

```diff
@@ -0,0 +1,192 @@
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.RequestPayer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.partition;
+import static com.google.common.collect.Multimaps.toMultimap;
+import static java.util.Objects.requireNonNull;
+
+final class S3FileSystem
+        implements TrinoFileSystem
+{
+    private final S3Client client;
+    private final S3Context context;
+    private final RequestPayer requestPayer;
+
+    public S3FileSystem(S3Client client, S3Context context)
+    {
+        this.client = requireNonNull(client, "client is null");
+        this.context = requireNonNull(context, "context is null");
+        this.requestPayer = context.requestPayer();
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location path)
+    {
+        return new S3InputFile(client, context, new S3Location(path), null);
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location path, long length)
+    {
+        return new S3InputFile(client, context, new S3Location(path), length);
+    }
+
+    @Override
+    public TrinoOutputFile newOutputFile(Location path)
+    {
+        return new S3OutputFile(client, context, new S3Location(path));
+    }
+
+    @Override
+    public void deleteFile(Location path)
+            throws IOException
+    {
+        path.verifyValidFileLocation();
+        var uri = new S3Location(path);
+        var request = DeleteObjectRequest.builder()
+                .requestPayer(requestPayer)
+                .key(uri.key())
+                .bucket(uri.bucket())
+                .build();
+
+        try {
+            client.deleteObject(request);
+        }
+        catch (SdkException e) {
+            throw new IOException("Failed to delete file: " + path, e);
+        }
+    }
+
+    @Override
+    public void deleteDirectory(Location path)
+            throws IOException
+    {
+        var iterator = listFiles(path);
+        while (iterator.hasNext()) {
+            List<Location> files = new ArrayList<>();
+            while ((files.size() < 1000) && iterator.hasNext()) {
+                files.add(iterator.next().location());
+            }
+            deleteFiles(files);
+        }
+    }
+
+    @Override
+    public void deleteFiles(Collection<Location> paths)
+            throws IOException
+    {
+        paths.forEach(Location::verifyValidFileLocation);
+
+        SetMultimap<String, String> bucketToKeys = paths.stream()
+                .map(S3Location::new)
+                .collect(toMultimap(S3Location::bucket, S3Location::key, HashMultimap::create));
+
+        Map<String, String> failures = new HashMap<>();
+
+        for (var entry : bucketToKeys.asMap().entrySet()) {
```

Comment:
I find all of the var's hard to read.  I'm constantly asking myself, what type is this thing